text/0045-docker-proxy.md

```diff
@@ -0,0 +1,176 @@
+  proxy, transparent proxy, and non-proxy environments
+* Have a uniform method to access proxy servers across Hyperledger Sawtooth
+  projects
+* Minimize Docker and script clutter with unnecessary tests and settings
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+In order to use the proposed containers, developers and users in proxied
+environments must set these proxy variables in their
+`$HOME/.docker/config.json` configuration file. The `config.json` file
+resides on the Docker host system, not in the Docker container filesystem.
+The relevant settings are:
+
+* `httpProxy` sets the proxy and port for unsecure web (HTTP) requests.
+  The setting is site-specific for your corporate proxy server. For example,
+  `"httpProxy": "http://proxy.foocorp.com:8080",`
+* `httpsProxy` sets the proxy and port for HTTPS (TLS) requests. For example,
+  `"httpsProxy": "http://proxy.foocorp.com:4443",`
+* `ftpProxy` sets the proxy for FTP file transfer (this is not common)
+* `noProxy` sets a comma-separated list of networks and hosts that are in
+  the internal network. For example,
+  `"noProxy": "192.168.0.0/16,172.16.0.0/12,10.0.0.0,localhost,127.0.0.0/8,::1,foocorp.com",`
+
+  Make sure you include Docker Compose internal network(s), which by default
+   are within `172.16.0.0/12` (use `netstat -rn` to list your networks)
+* Do not set `http_proxy` , `https_proxy` , and `no_proxy` or their upper
+  case equivalents. This method has been replaced with the `config.json`
+  settings.
+* Proxy support should be briefly mentioned in the README.md file or similar
+  file, along with the proxy variable(s) needed, such as `httpProxy`
+
+The following is a generic example of `$HOME/.docker/config.json` . The
+actual contents are site specific:
+
+```
+{
+ "proxies":
+ {
+   "default":
+   {
+     "httpProxy": "http://proxy.foocorp.com:8080",
+     "httpsProxy": "http://proxy.foocorp.com:4443",
+     "noProxy": "192.168.0.0/16,172.16.0.0/12,10.0.0.0,localhost,127.0.0.0/8,::1,foocorp.com"
+   }
+ }
+}
+````
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+All Docker files must be updated to conform as follows:
+* Commands `apt, apt-get, cargo, deb, go, npm, pip3, wget` and others require
+  no changes
+* `curl` requires no changes, but it is not usually installed. It needs to be
+  installed for containers that use it. For example,
+  `RUN apt-get update && apt-get install -yq curl `
+
+* `apt-key` : Instead of using `apt-key adv` to download an apt signing key,
+  use `curl` and `apt-key add -` . For example, instead of the following in a
+  `Dockerfile` (which no longer works):
+
+```
+RUN \
+ apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 \
+ --recv-keys 8AA7AF1F1091A5FD
+```
+
+Use the following, which works in proxy and non-proxy environments.
+Although not proxy-related, this example also has retry logic and uses
+two keyservers for robustness:
+
+```
+RUN \
```

Comment:
Since this RFC is going to become the reference document, I'd like to see this section be more concise and readable.

If we're adding retries, I think we can dispense with the backup keyserver.

Also, this doesn't work unless gnupg2 is installed.

Suggestion:
```
RUN apt-get update \
 && apt-get install -yq \
 curl \
 gnupg2

RUN curl --retry 3 -sSL \
    'https://keyserver.ubuntu.com/pks/lookup?op=get&search=0x8AA7AF1F1091A5FD' \
    | apt-key add -
```

Reply:
This is what I had in mind with my earlier comment about using a curl retry.

Reply:
I removed the alternate keyserver as suggested by others and kept the `--retry 3`
Retry logic and alternate keyservers are not really a proxy issue.

Reply:
Package gnupg2 is not needed and it works for me without it being installed.  I only see documentation in the package:

```
dpkg-query -L gnupg
/.
/usr
/usr/share
/usr/share/doc
/usr/share/doc/gnupg
/usr/share/doc/gnupg/DETAILS.gz
/usr/share/doc/gnupg/FAQ
/usr/share/doc/gnupg/HACKING.gz
/usr/share/doc/gnupg/OpenPGP.gz
/usr/share/doc/gnupg/README.Debian
/usr/share/doc/gnupg/README.gz
/usr/share/doc/gnupg/copyright
/usr/share/info
/usr/share/info/gnupg-card-architecture.png
/usr/share/info/gnupg-module-overview.png
/usr/share/info/gnupg.info-1.gz
/usr/share/info/gnupg.info-2.gz
/usr/share/info/gnupg.info.gz
/usr/share/doc/gnupg/NEWS.Debian.gz
/usr/share/doc/gnupg/NEWS.gz
/usr/share/doc/gnupg/THANKS.gz
/usr/share/doc/gnupg/TODO
/usr/share/doc/gnupg/changelog.Debian.gz
```

Reply:
After removing packages `gnupg` and documentation package `gnupg2`, both `apt-key adv` and `apt-key add` both still work.  See example below.

```
$ sudo apt remove gnupg gnupg2
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following packages will be REMOVED:
  gnupg gnupg2
0 upgraded, 0 newly installed, 2 to remove and 0 not upgraded.
After this operation, 423 kB disk space will be freed.
Do you want to continue? [Y/n] y
(Reading database ... 184249 files and directories currently installed.)
Removing gnupg2 (2.2.4-1ubuntu1.2) ...
Removing gnupg (2.2.4-1ubuntu1.2) ...
Processing triggers for man-db (2.8.3-2ubuntu0.1) ...
$ sudo apt-key del  44FC67F19B2466EA                              
OK
$ apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 44FC67F19B2466EA
Executing: /tmp/dano/apt-key-gpghome.yCD8U38W9A/gpg.1.sh --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 44FC67F19B2466EA
gpg: key 44FC67F19B2466EA: public key "Hyperledger Sawtooth Nightly Signing Key <distributedledger@intel.com>" imported
gpg: Total number processed: 1
gpg:               imported: 1
gpg: no writable keyring found: Not found
gpg: error reading '[stdin]': General error
gpg: import from '[stdin]' failed: General error
gpg: Total number processed: 0
$ sudo apt-key del  44FC67F19B2466EA                              
OK
$ curl --retry 3 -sSL 'https://keyserver.ubuntu.com/pks/lookup?op=get&search=0x44FC67F19B2466EA' | sudo apt-key add -                
OK
```